test/cpp/end2end/client_lb_end2end_test.cc

```diff
@@ -996,6 +1004,199 @@ TEST_F(ClientLbEnd2endTest, RoundRobinSingleReconnect) {
+        ::CreateLoadBalancingPolicy(delegate_policy_name.c_str(), args);
+    grpc_pollset_set_add_pollset_set(
+        delegate_->interested_parties(),
+        interested_parties());
+  }
+
+  void UpdateLocked(const grpc_channel_args& args) override {
+    delegate_->UpdateLocked(args);
+  }
+
+  bool PickLocked(PickState* pick, grpc_error** error) override {
+    return delegate_->PickLocked(pick, error);
+  }
+
+  void CancelPickLocked(PickState* pick, grpc_error* error) override {
+    delegate_->CancelPickLocked(pick, error);
+  }
+
+  void CancelMatchingPicksLocked(uint32_t initial_metadata_flags_mask,
+                                 uint32_t initial_metadata_flags_eq,
+                                 grpc_error* error) override {
+    delegate_->CancelMatchingPicksLocked(
+        initial_metadata_flags_mask,
+        initial_metadata_flags_eq,
+        error);
+  }
+
+  void NotifyOnStateChangeLocked(grpc_connectivity_state* state,
+                                 grpc_closure* closure) override {
+    delegate_->NotifyOnStateChangeLocked(state, closure);
+  }
+
+  grpc_connectivity_state CheckConnectivityLocked(
+      grpc_error** connectivity_error) override {
+    return delegate_->CheckConnectivityLocked(connectivity_error);
+  }
+
+  void HandOffPendingPicksLocked(LoadBalancingPolicy* new_policy) override {
+    delegate_->HandOffPendingPicksLocked(new_policy);
+  }
+
+  void ExitIdleLocked() override{
+    delegate_->ExitIdleLocked();
+  }
+
+  void ResetBackoffLocked() override {
+    delegate_->ResetBackoffLocked();
+  }
+
+  void FillChildRefsForChannelz(
+      grpc_core::channelz::ChildRefsList* child_subchannels,
+      grpc_core::channelz::ChildRefsList* ignored) override {
+    delegate_->FillChildRefsForChannelz(child_subchannels, ignored);
+  }
+
+ protected:
+  void ShutdownLocked() override {
+    // noop
+  }
+  Args args_;
+
+ private:
+  grpc_core::OrphanablePtr<LoadBalancingPolicy> delegate_;
+};
+
+class ClientLbInterceptTrailingMetadataTest : public ClientLbEnd2endTest {
+ protected:
+  void SetUp() override {
+    ClientLbEnd2endTest::SetUp();
+    grpc_core::LoadBalancingPolicyRegistry::Builder::
+        RegisterLoadBalancingPolicyFactory(
+            grpc_core::UniquePtr<grpc_core::LoadBalancingPolicyFactory>(
+                grpc_core::New<InterceptTrailingFactory>(this)));
+  }
+
+  void TearDown() override {
+    ClientLbEnd2endTest::TearDown();
+  }
+
+  class InterceptTrailingLb : public ForwardingLoadBalancingPolicy {
+   public:
+    InterceptTrailingLb(
+        const Args& args,
+        const std::string& delegate_lb_policy_name,
+        ClientLbInterceptTrailingMetadataTest* test)
+        : ForwardingLoadBalancingPolicy(args, delegate_lb_policy_name),
+        test_{test} {
+    }
+
+    bool PickLocked(PickState* pick, grpc_error** error) override {
+      bool ret = ForwardingLoadBalancingPolicy::PickLocked(pick, error);
+      // If these asserts fail, then we will need to add code to
```

Comment:
Note that these asserts aren't necessarily in the right place to catch this problem.  If the delegate LB policy returns asynchronously but doesn't set these fields until right before it invokes the completion callback, then they won't yet be set at this point, but there could still be a problem.

In order to handle that, I think we'll need to override the completion callback before invoking the delegate policy's `PickLocked()` method.  The callback we override will then set these fields before invoking the original completion callback.  And in this method, we only set these fields if the pick returns synchronously.

Alternatively, if we only plan to test with policies like pick_first that don't intercept trailing metadata themselves, then this is a non-issue.  But let's at least add a comment about it.